<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>R interface to the DIAMOND Sequence Aligner • rdiamond</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="R interface to the DIAMOND Sequence Aligner">
<meta property="og:description" content="R interface to the DIAMOND Sequence Aligner. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">rdiamond</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="rdiamond" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#rdiamond" class="anchor"></a>rdiamond</h1></div>
<div id="an-easy-to-use-framework-to-perform-massive-diamond-sequence-searches-with-r" class="section level2">
<h2 class="hasAnchor">
<a href="#an-easy-to-use-framework-to-perform-massive-diamond-sequence-searches-with-r" class="anchor"></a>An easy-to-use framework to perform massive <a href="http://www.diamondsearch.org/index.php">DIAMOND</a> sequence searches with R</h2>
<div id="motivation" class="section level3">
<h3 class="hasAnchor">
<a href="#motivation" class="anchor"></a>Motivation</h3>
<p><strong>Homepage</strong>: <a href="https://drostlab.github.io/rdiamond/" class="uri">https://drostlab.github.io/rdiamond/</a></p>
<p><a href="http://www.diamondsearch.org/index.php">DIAMOND</a> is a sequence aligner for protein and translated DNA searches, designed for high performance analysis of big sequence data. The key features are:</p>
<ul>
<li>Pairwise alignment of proteins and translated DNA at 100x-20,000x speed of BLAST.</li>
<li>Frameshift alignments for long read analysis.</li>
<li>Low resource requirements and suitable for running on standard desktops or laptops.</li>
<li>Various output formats, including BLAST pairwise, tabular and XML, as well as taxonomic classification.</li>
</ul>
<p>The <code>rdiamond</code> package provides easy-to-use interface functions between R and the standalone command line tool <a href="http://www.diamondsearch.org/index.php">DIAMOND</a>.</p>
<p><code>rdiamond</code> is designed to enable a new level of data-driven genomics research by providing the computational tools and data science standards needed to perform reproducible genomics research at scale.</p>
<p>The exponentially growing number of available sequences in biological databases revolutionizes the way modern life science research is conducted. Approximately hundred thousand genomic sequences spanning diverse species from the tree of life are currently publicly available and free to access. It is now possible to access and retrieve this data automatically using the R package <a href="https://github.com/ropensci/biomartr">biomartr</a> and the next step is to harness this wealth of sequence diversity to explore and detect novel patterns of evolvability, variation, and disease emergence using <code>rdiamond</code>.</p>
<p>The R package <a href="https://github.com/ropensci/biomartr">biomartr</a> <strong>solves the problem of retrieving this vast amount of biological sequence data</strong> in a standardized and computationally reproducible way and the <code>rdiamond</code> package aims to <strong>solve the problem of performing massive sequence searches on a tree-of-life scale</strong> in a standardized and computationally reproducible way.</p>
<p>Both packages, <code>biomartr</code> and <code>rdiamond</code> are designed to complement each other seamlessly to provide users with a tool set to automatically retrieve thousands of biological sequences (thousands of genomes, proteomes, annotations, etc) and to use these sequences to perform massive sequence searches with <a href="http://www.diamondsearch.org/index.php">DIAMOND</a> to extract novel patterns of similarity and divergence between large sets of species.</p>
</div>
<div id="install-rdiamond" class="section level3">
<h3 class="hasAnchor">
<a href="#install-rdiamond" class="anchor"></a>Install <code>rdiamond</code>
</h3>
</div>
<div id="for-linux-users" class="section level3">
<h3 class="hasAnchor">
<a href="#for-linux-users" class="anchor"></a>For Linux Users:</h3>
<p>Please install the <code>libpq-dev</code> library on you linux machine by typing into the terminal:</p>
<pre><code>sudo apt-get install libpq-dev</code></pre>
</div>
<div id="for-all-systems-install-rdiamond-by-typing" class="section level3">
<h3 class="hasAnchor">
<a href="#for-all-systems-install-rdiamond-by-typing" class="anchor"></a>For all systems install <code>rdiamond</code> by typing</h3>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co"># install Bioconductor</span>
<span class="co">if</span> (<span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, quietly = <span class="fl">TRUE</span>))
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw">BiocManager</span>::<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>()
<span class="co"># install Biostrings -&gt; see here for different Biostrings verions:</span>
<span class="co"># http://bioconductor.org/about/release-announcements/</span>
<span class="kw">BiocManager</span>::<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Biostrings"</span>))

<span class="co"># install.packages("devtools")</span>
<span class="co"># install the current version of rdiamond on your system</span>
<span class="kw">devtools</span>::<span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"drostlab/rdiamond"</span>, build_vignettes = <span class="fl">TRUE</span>, dependencies = <span class="fl">TRUE</span>)
</pre></div>
</div>
<div id="quick-start" class="section level3">
<h3 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h3>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="co"># run diamond assuming that the diamond executable is available</span>
<span class="co"># via the system path ('diamond_exec_path = NULL') and using</span>
<span class="co"># sensitivity_mode = "ultra-sensitive"</span>
<span class="kw">diamond_example</span> <span class="op">&lt;-</span> <span class="kw">rdiamond</span>::<span class="fu"><a href="reference/diamond_protein_to_protein.html">diamond_protein_to_protein</a></span>(
              query   = <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'seqs/qry_aa.fa'</span>, package = <span class="st">'rdiamond'</span>),
              subject = <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'seqs/sbj_aa.fa'</span>, package = <span class="st">'rdiamond'</span>),
              sensitivity_mode = <span class="st">"ultra-sensitive"</span>,
              output_path = <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span>(),
              db_import  = <span class="fl">FALSE</span>)

<span class="co"># look at DIAMOND results</span>
<span class="kw">diamond_example</span>
</pre></div>
<pre><code>Running diamond with 'diamond blastp  with query: /library/rdiamond/seqs/qry_aa.fa and subject: /library/rdiamond/seqs/sbj_aa.fa using 2 core(s) ...


Alignment Sensitivity Mode:  --ultra-sensitive and max-target-seqs: 500


Masking of low-complexity regions: TANTAN


diamond v2.0.4.142 (C) Max Planck Society for the Advancement of Science
Documentation, support and updates available at http://www.diamondsearch.org

#CPU threads: 4
Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)
Database input file: /library/rdiamond/seqs/sbj_aa.fa
Opening the database file...  [0.001s]
Loading sequences...  [0s]
Masking sequences...  [0.003s]
Writing sequences...  [0s]
Hashing sequences...  [0s]
Loading sequences...  [0s]
Writing trailer...  [0s]
Closing the input file...  [0s]
Closing the database file...  [0s]
Database hash = e7cd8e84df51b22dd27f3c01d5765fe1
Processed 20 sequences, 9444 letters.
Total time = 0.006s
diamond v2.0.4.142 (C) Max Planck Society for the Advancement of Science
Documentation, support and updates available at http://www.diamondsearch.org

#CPU threads: 2
Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)
Temporary directory: /var/folders/yn/mgwl8_b56hz4v2c2vlfxj07w00076j/T//RtmpUCSsbL
Opening the database...  [0.002s]
#Target sequences to report alignments for: 500
Reference = /var/folders/yn/mgwl8_b56hz4v2c2vlfxj07w00076j/T//RtmpUCSsbL/sbj_aa.fa.dmnd
Sequences = 20
Letters = 9444
Block size = 400000000
Opening the input file...  [0s]
Opening the output file...  [0s]
Loading query sequences...  [0s]
Masking queries...  [0.002s]
Building query seed set... Algorithm: Double-indexed
 [0s]
Building query histograms...  [0.008s]
Allocating buffers...  [0s]
Loading reference sequences...  [0s]
Masking reference...  [0.001s]
Initializing temporary storage...  [0.013s]
Building reference histograms...  [0.007s]
Allocating buffers...  [0s]
Processing query block 1, reference block 1/1, shape 1/64.
Building reference seed array...  [0s]
Building query seed array...  [0s]
Computing hash join...  [0.001s]
Building seed filter...  [0.001s]
Searching alignments...  [0.002s]

... 


Building reference seed array...  [0s]
Building query seed array...  [0s]
Computing hash join...  [0s]
Building seed filter...  [0.033s]
Searching alignments...  [0.001s]
Deallocating buffers...  [0s]
Clearing query masking...  [0s]
Computing alignments...  [0.065s]
Deallocating reference...  [0s]
Loading reference sequences...  [0s]
Deallocating buffers...  [0s]
Deallocating queries...  [0s]
Loading query sequences...  [0s]
Closing the input file...  [0s]
Closing the output file...  [0s]
Closing the database file...  [0s]
Deallocating taxonomy...  [0s]
Total time = 0.797s
Reported 20 pairwise alignments, 20 HSPs.
20 queries aligned.


DIAMOND search finished successfully! 


The DIAMOND output file was imported into the running R session. The DIAMOND output file has been stored at: /var/folders/RtmpUCSsbL/qry_aa_sbj_aa_blastp_eval_0.001.blast_tbl</code></pre>
<pre><code>A tibble: 20 x 20
   query_id subject_id perc_identity num_ident_match…
   &lt;chr&gt;    &lt;chr&gt;              &lt;dbl&gt;            &lt;int&gt;
 1 333554|… AT1G01010…          73.2              347
 2 470181|… AT1G01020…          91.1              224
 3 470180|… AT1G01030…          93.3              335
 4 333551|… AT1G01040…          93.4             1840
 5 909874|… AT1G01050…         100                213
 6 470177|… AT1G01060…          87.5              567
 7 918864|… AT1G01070…          92.6              339
 8 909871|… AT1G01080…          89.3              268
 9 470171|… AT1G01090…          96.8              420
10 333544|… AT1G01110…          87.7              463
11 918858|… AT1G01120…          99.2              525
12 470161|… AT1G01140…          98.5              446
13 918855|… AT1G01150…          72.6              207
14 918854|… AT1G01160…          78.8              141
15 311317|… AT1G01170…          85.6               83
16 909860|… AT1G01180…          92.6              287
17 311315|… AT1G01190…          94.2              502
18 470156|… AT1G01200…          95.8              228
19 311313|… AT1G01210…          95.3              102
20 470155|… AT1G01220…          96.5             1019
# … with 16 more variables: alig_length &lt;int&gt;,
#   mismatches &lt;int&gt;, gap_openings &lt;int&gt;, n_gaps &lt;int&gt;,
#   pos_match &lt;int&gt;, ppos &lt;dbl&gt;, q_start &lt;int&gt;,
#   q_end &lt;int&gt;, q_len &lt;int&gt;, qcovhsp &lt;dbl&gt;, s_start &lt;int&gt;,
#   s_end &lt;int&gt;, s_len &lt;int&gt;, evalue &lt;dbl&gt;,
#   bit_score &lt;dbl&gt;, score_raw &lt;dbl&gt;</code></pre>
</div>
</div>
<div id="discussions-and-bug-reports" class="section level2">
<h2 class="hasAnchor">
<a href="#discussions-and-bug-reports" class="anchor"></a>Discussions and Bug Reports</h2>
<p>I would be very happy to learn more about potential improvements of the concepts and functions provided in this package.</p>
<p>Furthermore, in case you find some bugs, need additional (more flexible) functionality of parts of this package, or want to contribute to this project please let me know:</p>
<p><a href="https://github.com/drostlab/rdiamond/issues" class="uri">https://github.com/drostlab/rdiamond/issues</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Hajk-Georg Drost <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-1567-306X" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
